### PR TITLE
Attempt at pi cow ap

### PR DESCRIPTION
**Pull request by bill88t**
This is an attempt to implement access point functionallity for the pi cow.
The access point itself works.

However:
- stop_ap does not work
- You can connect to an access point successfully even when the ap is still running..
  (This is due to CYW43_LINK_UP)

### CODE REVIEW

**Comment by anecdata** [comment]
> You can connect to an access point successfully even when the ap is still running.

This is normal in `espressif` where you can have independent control of whether STA and / or AP are up.

**Comment by bill88t** [comment]
Oh sorry I didn't phrase it very well.
With the ap open, you can connect to a network and it will connect 
But any connections will fail. cyw43 will start crying debug messages.
I just have to fix the stop_ap and call it before a connection.

**Comment by anecdata** [comment]
OK, sorry, missed that this was in draft.

**Comment by bill88t** [comment]
Updated my local branch, cherry-picked the commits from [paul-1's fork](https://github.com/paul-1/circuitpython/tree/picow-AP-mode-with-DHCP-server), thanks @paul-1!
And I am now giving it another try.
As soon as I get the disconnection working, we should be able to merge this.

**Comment by tannewt** [comment]
> Perhaps before wifi init we should use getenv to look at see if the user has a valid country code defined in the .env file?

That sounds like a good use for `.env`! :+1:

**Comment by bill88t** [comment]
I cannot get the ap to stop no matter what I do..
This is due to `cyw43_wifi_leave` being borked and doing nothing.

Looking at the respective micropython [code](https://github.com/micropython/micropython/blob/master/drivers/cyw43/cyw43_ctrl.c) we can see, they haven't implemented it either.

I will continue implementing the AP, without `stop_ap`.
`stop_ap` will simply raise an appropriate error. Feel free to suggest error messages.
What I propose is `NotImplementedError: AP cannot be disconnected.`

**Comment by anecdata** [comment]
What happens if `start_ap()` is called a second time, with different parameters such as `ssid` or `password`?

Do we have any API to access `cyw43_deinit()` to shut it all down and start over?

**Comment by bill88t** [comment]
> What happens if `start_ap()` is called a second time, with different parameters such as `ssid` or `password`?

Stays on original ssid and accepts no connections.
This applies for same ssid, different passwd
different ssid, same passwd
different ssid, different passwd

> Do we have any API to access `cyw43_deinit()` to shut it all down and start over (short of a CircuitPython reload)?

Aren't some pins controlled from the wifi chip?
It would be pretty problematic resetting.
Pretty sure this is the reason it hasn't been implemented so far.

Note: I have not touched any of the early-boot cyw43 code.
I am absolutely do not know a thing.

**Comment by bill88t** [comment]
Now, `wifi.radio.start_ap`, `wifi.radio.stop_ap`, `wifi.radio.connect` and `wifi.radio.stop_station` are ready.
I only now need to do the getters for the ap.

The error messages I added are very much suboptimal. I would very much appreciate some suggestions.

**Comment by paul-1** [comment]
TBH, I'm not sure there is significant value top stop the AP during a running program.  Changing the AP SSID/password, then restarting the code worked fine when I previously tested.   I think there are things that just requires the radio to be hard power cycled.  For example, once the AP is started the channel assignment will not change without a hard power cycle.

I looked at the error messages.  "cannot be stopped" has the connotation that something is preventing it from stopping.  IMO, "Stopping AP is not supported" is a better choice.

**Comment by anecdata** [comment]
I have an application (on ESP32-S*) that sets up different SSIDs at different times during a run.

We are (hopefully) trying to have a single consistent `wifi` API for both espressif and raspberrypi, but unfortunately the raspberrypi underpinnings are less mature / less granular, so we'll have some limitations (for now?)

**Comment by bill88t** [comment]
Hopefully as cyw43 is updated, this stuff will be fixed.
However now we are stuck with what we got.
`microcontroller.reset()` is still an option.

**Comment by paul-1** [comment]
"Significant Value" was a poor choice of words, on my part. I agree with common api within CP. Since we have to wait for SDK/firmware improvements, I propose we just stick with the current state of the AP development. Cleanup and merge for 8.0.0, and add an open issue or milestone to re-evaluate for future releases.

**Comment by bill88t** [comment]
Actually `ap_info` is for "the access point we are connected to". So there is nothing more to implement.
Ready for review.

**Review by anecdata — COMMENTED** [review]
Differences from `espressif`, from above and from testing:
- Station and AP can’t operate simultaneously
-- if device AP is started, device Station can't connect to an external AP (~~RuntimeError: Already in access point mode.~~ `RuntimeError: Wifi is in access point mode.`)
-- If device Station is connected to an external AP, device AP can't be started (~~RuntimeError: Already connected to station.~~ `RuntimeError: Wifi is in station mode.`)
- once an AP is started, it can't be stopped without a power cycle or `microcontroller.reset()`
- AP channel (_edit:_ and other AP settings) can only be set once after a power cycle or `microcontroller.reset()`
- `wifi.radio.enabled` is not implemented (`AttributeError: 'Radio' object has no attribute 'enabled'`), so to disconnect from an external AP, use `wifi.radio.stop_station()` --> _EDIT: this has now been implemented_ --> _EDIT 2:_ `.enabled` isn't implemented, but is checked for certain logic.

Thanks!

_Addendum:_

- wifi country is set to global wifi policy

File: ports/raspberrypi/common-hal/wifi/Radio.c
Comment (bill88t) [reply ports/raspberrypi/common-hal/wifi/Radio.c]: Changed the messages up a bit.
These should be a lot simplier.

**Comment by bill88t** [comment]
> * once an AP is started, it can't be stopped

Or even have its settings changed.
You get one shot. :sweat_smile:

**Review by anecdata — DISMISSED** [review]
Everything I've tried is working as expected. Just ran an `adafruit_httpserver` on the wifi AP, and connected with a CircuitPython client and got served. I'm going to approve this and defer to one of the core devs to decide to merge. It would be nice to get this out there for more folks to poke at it :-)

**Comment by bill88t** [comment]
```
1230984 bytes used, 337784 bytes free in flash firmware space out of 1568768 bytes (1532.0kB).
67784 bytes used, 194360 bytes free in ram for stack and heap out of 262144 bytes (256.0kB).
```
Just in case anyone cares about free board storage.

**Comment by dhalbert** [comment]
@jepler Could you look this over before we merge? Thanks.

@bill88t There are now merge conflicts that need to be resolved.

**Comment by bill88t** [comment]
Built and tested locally.
```
1259888 bytes used, 308880 bytes free in flash firmware space out of 1568768 bytes (1532.0kB).
69044 bytes used, 193100 bytes free in ram for stack and heap out of 262144 bytes (256.0kB).
```

One thing I noticed however, these messages pop up whenever a device connects.
`DHCPS: client connected: MAC=8a:ab:19:62:52:ad IP=192.168.4.16`
The build is not a debug build. I'm pretty sure they should be disabled on regular builds.

**Comment by bill88t** [comment]
`shared/netutils/dhcpserver.c:268`
Why isn't there 'some' condition? This needs to be silenced in some way..

**Comment by jepler** [comment]
perhaps
```patch
diff --git a/shared/netutils/dhcpserver.c b/shared/netutils/dhcpserver.c
index a61501c93c..d396a2ba56 100644
--- a/shared/netutils/dhcpserver.c
+++ b/shared/netutils/dhcpserver.c
@@ -265,9 +265,9 @@ static void dhcp_server_process(void *arg, struct udp_pcb *upcb, struct pbuf *p,
             d->lease[yi].expiry = (mp_hal_ticks_ms() + DEFAULT_LEASE_TIME_S * 1000) >> 16;
             dhcp_msg.yiaddr[3] = DHCPS_BASE_IP + yi;
             opt_write_u8(&opt, DHCP_OPT_MSG_TYPE, DHCPACK);
-            printf("DHCPS: client connected: MAC=%02x:%02x:%02x:%02x:%02x:%02x IP=%u.%u.%u.%u\n",
+            LWIP_DEBUGF(DHCP_DEBUG, ("DHCPS: client connected: MAC=%02x:%02x:%02x:%02x:%02x:%02x IP=%u.%u.%u.%u\n",
                 dhcp_msg.chaddr[0], dhcp_msg.chaddr[1], dhcp_msg.chaddr[2], dhcp_msg.chaddr[3], dhcp_msg.chaddr[4], dhcp_msg.chaddr[5],
-                dhcp_msg.yiaddr[0], dhcp_msg.yiaddr[1], dhcp_msg.yiaddr[2], dhcp_msg.yiaddr[3]);
+                dhcp_msg.yiaddr[0], dhcp_msg.yiaddr[1], dhcp_msg.yiaddr[2], dhcp_msg.yiaddr[3]));
             break;
         }
 
```

**Comment by bill88t** [comment]
Oh I thought it was a submodule. Thankfully it is not.
Imma test it just in case (w & wo/ debug), and we are good to go.

File: ports/raspberrypi/supervisor/port.c
Comment (jepler) [inline ports/raspberrypi/supervisor/port.c:148]: I filed https://github.com/adafruit/circuitpython/issues/7285 about this as I could not find an existing issue

**Review by jepler — APPROVED** [review]
Thanks for working on this! No testing performed.

File: ports/raspberrypi/common-hal/wifi/Radio.c
Comment (bill88t) [reply ports/raspberrypi/common-hal/wifi/Radio.c]: Ok, I improved upon the old comment.

File: ports/raspberrypi/common-hal/wifi/Radio.c
Comment (bill88t) [reply ports/raspberrypi/common-hal/wifi/Radio.c:200]: I am pretty sure there isn't.
But I do not hold the technical knowledge to push such an issue ahead.
I would really prefer someone else more knowledgeable handled this.
However should nobody step up, I will do it.

**Review by jepler — APPROVED** [review]
Thank you again!

File: shared/netutils/dhcpserver.h
Comment (jepler) [inline shared/netutils/dhcpserver.h:35]: won't this turn the debug messages back on? is that what you intended?

**Comment by bill88t** [comment]
Now when `DEBUG=1`, dhcps should announce connections.
This debugging info is useful and shouldn't be barriered behind the header file.
Testing, and I am done with this pr.

File: ports/raspberrypi/common-hal/wifi/Radio.c
Comment (jepler) [reply ports/raspberrypi/common-hal/wifi/Radio.c:200]: I filed https://github.com/georgerobotics/cyw43-driver/issues/47

File: shared/netutils/dhcpserver.h
Comment (bill88t) [reply shared/netutils/dhcpserver.h:35]: When `DEBUG=0` the user shouldn't see them.
When `DEBUG=1` you probably want them.

Comment (bill88t) [reply shared/netutils/dhcpserver.h:35]: It builds (w & wo/ debug) but doesn't work though, and I am not sure why.

**Comment by bill88t** [comment]
Idk how to fix it, and it really doesn't matter.
I will continue testing locally, and follow up with more pr's on picow ap.
As much as this one is concerned, done.

**Review by jepler — APPROVED** [review]
Thanks! I do not have a strong opinion about that dhcp debug message, it can be fixed in the future with a fresh PR.

**Pull request merged**
